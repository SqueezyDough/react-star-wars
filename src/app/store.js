import { createStore } from 'redux'
import { persistStore, persistReducer } from 'redux-persist'
import storage from 'redux-persist/lib/storage'
import resourceReducer from '../features/resource/resourceSlice'

const persistConfig = {
  key: 'root',
  storage
}

const persistedReducer = persistReducer(persistConfig, resourceReducer)

export const store = createStore(
  persistedReducer,
  window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__()
)
export const persistor = persistStore(store)
