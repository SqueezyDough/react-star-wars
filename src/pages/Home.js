import React, { useState, useEffect } from 'react'
import { useSelector } from 'react-redux'
import { Header, CardsList, Selection } from '../components/index'
import { Api } from '../modules/index'
import ErrorBoundary from '../ErrorBoundary'
import { initState } from '../features/resource/resourceSlice'

function Home () {
  const { selectedOption } = useSelector(state => state)

  // manages all options
  const [options, setOptions] = useState([])

  // manages all characters from that resource
  const [characters, setCharacters] = useState([])

  // loading state
  const [loadingState, setLoadingState] = useState({
    start: true,
    loaders: 8
  })

  // fetch all options once
  useEffect(() => {
    Api.fetchData(process.env.REACT_APP_API)
      .then(res => Object.entries(res))
      .then(options => {
        initState(options.map(([key]) => key))
        setOptions(() => options.map(option => {
          return {
            value: { name: option[0], url: option[1] }
          }
        }))
      })
      .catch(() => {})
  }, [])

  // whenever our resource changes, fetch characters
  useEffect(() => {
    if (selectedOption) {
      setLoadingState(prevState => {
        const updatedState = {
          start: true,
          loaders: prevState.loaders
        }
        return { ...prevState, ...updatedState }
      })

      Api.fetchData(`${process.env.REACT_APP_API}${selectedOption}`)
        .then(res => setCharacters(res))
        .catch(() => {})
        .finally(() => {
          setLoadingState(prevState => {
            const updatedState = {
              start: false,
              loaders: prevState.loaders
            }
            return { ...prevState, ...updatedState }
          })
        })
    }
  }, [selectedOption])

  return (
    <main className='content-wrapper'>
      <Header />
      <span>What are you looking for?</span>

      {/* stupid comment */}

      <ErrorBoundary>
        <Selection options={options} />
      </ErrorBoundary>
      <ErrorBoundary>
        <CardsList characters={characters.results} loadingState={loadingState} />
      </ErrorBoundary>
    </main>
  )
}

export default Home
