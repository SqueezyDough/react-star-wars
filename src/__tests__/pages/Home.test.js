import React from 'react'
import renderer from 'react-test-renderer'
import Home from '../../pages/Home'
import { Provider } from 'react-redux'
import { store, persistor } from '../../app/store'
import { PersistGate } from 'redux-persist/integration/react'

it('renders correctly', () => {
  const tree = renderer
    .create(
      <Provider store={store}>
        <PersistGate loading={null} persistor={persistor}>
          <Home />
        </PersistGate>
      </Provider>
    )
    .toJSON()

  expect(tree).toMatchSnapshot()
})
