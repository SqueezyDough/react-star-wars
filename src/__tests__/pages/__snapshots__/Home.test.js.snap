// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
.c0 {
  display: block;
  margin: 0;
  padding: 3rem 0;
  font-weight: bold;
  line-height: 7rem;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c0 span:last-of-type {
  display: block;
}

.c0 span:last-of-type::after {
  content: '.';
}

.c6 {
  display: grid;
  grid-template-columns: 1fr 1fr 1fr;
  -webkit-column-gap: 1rem;
  column-gap: 1rem;
  row-gap: 1rem;
  margin-top: 5rem;
  padding: 1rem 0;
}

.c1 {
  display: inline;
}

.c1 select {
  margin: 3rem .5rem 0;
  padding: .5rem 1rem;
  border: none;
}

.c3 {
  position: absolute;
  width: 100%;
  min-height: 60vh;
  z-index: 1;
}

.c4 {
  display: grid;
  grid-template-columns: 1fr 1fr 1fr;
  -webkit-column-gap: 1rem;
  column-gap: 1rem;
  row-gap: 1rem;
  margin-top: 5rem;
  padding: 1rem 0;
  overflow: hidden;
}

.c2 {
  position: relative;
  overflow: hidden;
}

.c2 ~ section {
  display: none;
}

.c2.done ~ section {
  display: block;
}

.c2 .load-image {
  display: block;
  height: 100%;
  overflow: hidden;
  -webkit-transform: skewX(30deg);
  -ms-transform: skewX(30deg);
  transform: skewX(30deg);
  -webkit-animation: skew .5s forwards;
  animation: skew .5s forwards;
  -webkit-animation-delay: .5s;
  animation-delay: .5s;
}

.c2 .load-image img {
  display: block;
  object-fit: cover;
  height: 100%;
  position: absolute;
  left: -1000px;
  -webkit-animation: slide 1s forwards;
  animation: slide 1s forwards;
}

.c5 {
  -webkit-transition: background;
  transition: background;
  text-align: center;
  -webkit-animation: pulse 1s infinite alternate;
  animation: pulse 1s infinite alternate;
}

.c5:first-of-type {
  grid-column: 2;
}

.c5 .card {
  padding: 10%;
}

.c5 .card__short-title {
  font-weight: bold;
}

<main
  className="content-wrapper"
>
  <a
    className="c0"
    href="/"
  >
    <span>
      Star Wars
    </span>
    <span>
      Universe
    </span>
  </a>
  <span>
    What are you looking for?
  </span>
  <div
    className="c1"
  >
    <select
      onChange={[Function]}
      value=""
    />
  </div>
  <section
    className="c2 "
  >
    <div
      className="c3"
    />
    <div
      className="c4"
    >
      <div
        className="load-image loader"
      >
        <img
          alt=""
          src={"http://localhost/sw-loading.gif"}
        />
      </div>
      <div
        className="c5 loader"
      >
        <article
          className="card"
        >
          <span
            className="card__short-title"
          >
            L
          </span>
          <h1
            className="card__title"
          >
            Loading...
          </h1>
        </article>
      </div>
      <div
        className="c5 loader"
      >
        <article
          className="card"
        >
          <span
            className="card__short-title"
          >
            L
          </span>
          <h1
            className="card__title"
          >
            Loading...
          </h1>
        </article>
      </div>
      <div
        className="c5 loader"
      >
        <article
          className="card"
        >
          <span
            className="card__short-title"
          >
            L
          </span>
          <h1
            className="card__title"
          >
            Loading...
          </h1>
        </article>
      </div>
      <div
        className="c5 loader"
      >
        <article
          className="card"
        >
          <span
            className="card__short-title"
          >
            L
          </span>
          <h1
            className="card__title"
          >
            Loading...
          </h1>
        </article>
      </div>
      <div
        className="c5 loader"
      >
        <article
          className="card"
        >
          <span
            className="card__short-title"
          >
            L
          </span>
          <h1
            className="card__title"
          >
            Loading...
          </h1>
        </article>
      </div>
      <div
        className="c5 loader"
      >
        <article
          className="card"
        >
          <span
            className="card__short-title"
          >
            L
          </span>
          <h1
            className="card__title"
          >
            Loading...
          </h1>
        </article>
      </div>
      <div
        className="c5 loader"
      >
        <article
          className="card"
        >
          <span
            className="card__short-title"
          >
            L
          </span>
          <h1
            className="card__title"
          >
            Loading...
          </h1>
        </article>
      </div>
      <div
        className="c5 loader"
      >
        <article
          className="card"
        >
          <span
            className="card__short-title"
          >
            L
          </span>
          <h1
            className="card__title"
          >
            Loading...
          </h1>
        </article>
      </div>
    </div>
  </section>
  <section>
    <div
      className="c6"
    />
  </section>
</main>
`;
